lightningd/pay.c

```diff
@@ -551,6 +575,11 @@ void payment_failed(struct lightningd *ld, const struct htlc_out *hout,
 		pay_errcode = PAY_UNPARSEABLE_ONION;
 		fail = NULL;
 		failmsg = NULL;
+	} else if (hout->failcode) {
```

Comment:
Whut?  How did we get a `failcode` here if the onion was malformed?  If we got a `failcode` from the direct peer, should not we consider it as a local failure and the `localfail` was set?

Reply:
The erring node was not able to parse the onion, but it was able to return an error (it knows it has the correct shared-secret since the HMAC test succeeds, but the parsing of the payload failed). This is `PERM|22 (invalid_onion_payload)` if I'm not mistaken, not one of the `BADONION|xyz` failcodes.

Reply:
If so, then the succeeding "Must be remote fail." case should cover it.  So it is this entire case which confuses me.

Reply:
localfail is a bit of a misnomer.  It means *lightningd* gave an error (eg. can't find next hop).  *channeld* returning an error goes down the "remote fail" path.

Let me add a comment: see fixup patch at end.